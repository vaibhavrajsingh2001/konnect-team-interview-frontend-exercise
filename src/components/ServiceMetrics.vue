<template>
  <div class="service-metrics">
    <div
      v-for="item in metricList"
      :key="item.label"
      class="service-metric-item"
    >
      <span
        class="ellipse"
        :class="item.ellipseColor"
      />
      {{ item.value }} <span class="service-metric-item-label">{{ item.label }}</span>
    </div>
  </div>
</template>
<script lang="ts">
import { humanizeNumber } from '@/util'
import { defineComponent } from 'vue'

enum EllipseColor {
  PASS = 'service-metric-pass',
  WARN = 'service-metric-warn',
  FAIL = 'service-metric-fail',
  NEUTRAL = 'service-metric-neutral'
}

interface MetricItem {
  label: string
  value: string
  ellipseColor: EllipseColor
}

export default defineComponent({
  name: 'ServiceMetrics',
  props: {
    latency: {
      type: Number,
      required: true,
    },
    uptime: {
      type: Number,
      required: true,
    },
    requests: {
      type: Number,
      required: true,
    },
    errors: {
      type: Number,
      required: true,
    },
  },
  computed: {
    metricList(): Array<MetricItem> {
      return [
        {
          label: 'latency',
          value: `${this.latency}ms`,
          ellipseColor: this.latency > 1 ? EllipseColor.FAIL : this.latency > 0.5 ? EllipseColor.WARN : EllipseColor.PASS,
        },
        {
          label: 'uptime',
          value: `${this.decimalToPercentage(this.uptime)}%`,
          ellipseColor: this.uptime < 0.7 ? EllipseColor.FAIL : this.uptime < 0.9 ? EllipseColor.WARN : EllipseColor.PASS,
        },
        {
          label: 'errors',
          value: `${this.decimalToPercentage(this.errors)}%`,
          ellipseColor: this.errors > 0.3 ? EllipseColor.FAIL : this.errors > 0.1 ? EllipseColor.WARN : EllipseColor.PASS,
        },
        {
          label: 'requests',
          value: this.humanizeNumber(this.requests),
          // Requests count can't be evaluated with a simple pass/warn/fail
          ellipseColor: EllipseColor.NEUTRAL,
        },
      ]
    },
  },
  methods: {
    humanizeNumber,
    decimalToPercentage(candidate: number) {
      return (candidate * 100).toFixed(2)
    },
  },
})
</script>

<style lang="scss">
.service-metrics {
  color: #3C4557;
  font-size: 1.2rem;
  font-weight: 600;
}

.service-metric-item {
  align-items: center;
  display: flex;
  gap: 0.9rem;
  margin-bottom: 0.5rem;

  .ellipse {
    background-color: #4CAF50;
    border-radius: 50%;
    display: inline-block;
    height: 0.6rem;
    width: 0.6rem;
  }

  // Colors for ellipses
  .service-metric-pass {
    background-color: #42D782;
  }

  .service-metric-fail {
    background-color: #d6435b;
  }

  .service-metric-warn {
    background-color: #FABE5F;
  }

  .service-metric-neutral {
    background-color: #ADB6C8;
  }

  .service-metric-item-label {
    color: #707888;
  }
}
</style>
