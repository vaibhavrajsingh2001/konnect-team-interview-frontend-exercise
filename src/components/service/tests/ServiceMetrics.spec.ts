import { renderToString } from '@vue/test-utils'
import { it, describe, expect } from 'vitest'
import ServiceMetrics from '@/components/service/ServiceMetrics.vue'

describe('ServiceMetrics', () => {
  it('renders correctly for all passing metrics', async () => {
    const wrapper = await renderToString(ServiceMetrics, {
      props: {
        latency: 0.42,
        uptime: 0.9607,
        requests: 31067,
        errors: 0.0724,
      }
      ,
    })
    expect(wrapper).toMatchSnapshot()
  })

  it('renders correctly for all failing metrics', async () => {
    const wrapper = await renderToString(ServiceMetrics, {
      props: {
        latency: 1.62,
        uptime: 0.6607,
        requests: 31067,
        errors: 0.4724,
      }
      ,
    })
    expect(wrapper).toMatchSnapshot()
  })

  it('renders correctly for all warning metrics', async () => {
    const wrapper = await renderToString(ServiceMetrics, {
      props: {
        latency: 0.6,
        uptime: 0.8,
        requests: 31067,
        errors: 0.2,
      }
      ,
    })
    expect(wrapper).toMatchSnapshot()
  })
})
