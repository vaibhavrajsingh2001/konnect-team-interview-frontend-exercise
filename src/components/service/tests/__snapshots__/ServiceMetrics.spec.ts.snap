// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ServiceMetrics > renders correctly for all failing metrics 1`] = `"<div class="service-metrics"><!--[--><div class="service-metric-item"><span class="ellipse service-metric-fail"></span> 1.62ms <span class="service-metric-item-label">latency</span></div><div class="service-metric-item"><span class="ellipse service-metric-fail"></span> 66.07% <span class="service-metric-item-label">uptime</span></div><div class="service-metric-item"><span class="ellipse service-metric-fail"></span> 47.24% <span class="service-metric-item-label">errors</span></div><div class="service-metric-item"><span class="ellipse service-metric-neutral"></span> 31.1k <span class="service-metric-item-label">requests</span></div><!--]--></div>"`;

exports[`ServiceMetrics > renders correctly for all passing metrics 1`] = `"<div class="service-metrics"><!--[--><div class="service-metric-item"><span class="ellipse service-metric-pass"></span> 0.42ms <span class="service-metric-item-label">latency</span></div><div class="service-metric-item"><span class="ellipse service-metric-pass"></span> 96.07% <span class="service-metric-item-label">uptime</span></div><div class="service-metric-item"><span class="ellipse service-metric-pass"></span> 7.24% <span class="service-metric-item-label">errors</span></div><div class="service-metric-item"><span class="ellipse service-metric-neutral"></span> 31.1k <span class="service-metric-item-label">requests</span></div><!--]--></div>"`;

exports[`ServiceMetrics > renders correctly for all warning metrics 1`] = `"<div class="service-metrics"><!--[--><div class="service-metric-item"><span class="ellipse service-metric-warn"></span> 0.6ms <span class="service-metric-item-label">latency</span></div><div class="service-metric-item"><span class="ellipse service-metric-warn"></span> 80.00% <span class="service-metric-item-label">uptime</span></div><div class="service-metric-item"><span class="ellipse service-metric-warn"></span> 20.00% <span class="service-metric-item-label">errors</span></div><div class="service-metric-item"><span class="ellipse service-metric-neutral"></span> 31.1k <span class="service-metric-item-label">requests</span></div><!--]--></div>"`;
